metamorphic: assignment to entry in nil map

I found this failure case with `TestMetaTwoInstance`:

```
Init(2 /* dbs */, 45 /* batches */, 67 /* iters */, 48 /* snapshots */)
db1.RangeKeySet("a", "c", "@10", "")
db1.Flush()
iter64 = db1.NewIter("", "", 2 /* key types */, 0, 0, false /* use L6 filters */, "" /* masking suffix */)
iter64.SeekGE("b", "")
db1.Close()
```
OPTIONS: https://gist.github.com/RaduBerinde/b2bb75ea16b178d6881edfe66d547935

```
=== RUN   TestMetaTwoInstance
// INFO: [JOB 1] MANIFEST created 000001
// INFO: [JOB 1] WAL created 000002
// db1.Open() <nil>
// INFO: [JOB 1] MANIFEST created 000001
// INFO: [JOB 2] all initial table stats loaded
// INFO: [JOB 1] WAL created 000002
// db2.Open() <nil>
Init(2 /* dbs */, 45 /* batches */, 67 /* iters */, 48 /* snapshots */) #0
// INFO: [JOB 2] all initial table stats loaded
db1.RangeKeySet("a", "c", "@10", "") // <nil> #1
// INFO: [JOB 3] WAL created 000004
// INFO: [JOB 4] flushing 1 memtable (43B) to L0
// INFO: [JOB 4] flushing: sstable created 000005
// INFO: [JOB 4] flushed 1 memtable (43B) to L0 [000005] (928B), in 0.0s (0.0s total), output rate 3.0MB/s
db1.Flush() // <nil> #2
iter64 = db1.NewIter("", "", 2 /* key types */, 0, 0, false /* use L6 filters */, "" /* masking suffix */) // <nil> #3
// INFO: [JOB 4] WAL deleted 000002
db1.Close() // leaked iterators: current
0.0:
  000005:[a#10,RANGEKEYSET-c#inf,RANGEKEYSET] #5
panic: assignment to entry in nil map

goroutine 58 [running]:
github.com/cockroachdb/pebble.(*tableCacheShard).addNode(0x1400027a000, 0x1400029a570, 0x1400020e488)
        /Users/radu/go/src/github.com/cockroachdb/pebble/table_cache.go:897 +0x8c
github.com/cockroachdb/pebble.(*tableCacheShard).findNodeInternal(0x1400027a000, 0x14000192a80, 0x1400020e488)
        /Users/radu/go/src/github.com/cockroachdb/pebble/table_cache.go:828 +0x3b4
github.com/cockroachdb/pebble.(*tableCacheShard).findNode(...)
        /Users/radu/go/src/github.com/cockroachdb/pebble/table_cache.go:778
github.com/cockroachdb/pebble.(*tableCacheShard).newRangeKeyIter(0x1400027a000, 0x14000192a80, {{0x0?, 0x1?, 0x140002b35d8?}}, 0x1400020e488)
        /Users/radu/go/src/github.com/cockroachdb/pebble/table_cache.go:596 +0x58
github.com/cockroachdb/pebble.(*tableCacheContainer).newRangeKeyIter(0x140002b3618?, 0x100cf50b0?, {{0x0?, 0x0?, 0x1?}})
        /Users/radu/go/src/github.com/cockroachdb/pebble/table_cache.go:150 +0x4c
github.com/cockroachdb/pebble/internal/keyspan.(*LevelIter).loadFile(0x140002cc778, 0x140002aa050?, 0x14000240250?)
        /Users/radu/go/src/github.com/cockroachdb/pebble/internal/keyspan/level_iter.go:182 +0x150
github.com/cockroachdb/pebble/internal/keyspan.(*LevelIter).SeekLT(0x140002cc778, {0x14000240250, 0x1, 0x8})
        /Users/radu/go/src/github.com/cockroachdb/pebble/internal/keyspan/level_iter.go:284 +0x2cc
github.com/cockroachdb/pebble/internal/keyspan.(*MergingIter).SeekGE(0x140002cc2e8, {0x14000240250, 0x1, 0x8})
        /Users/radu/go/src/github.com/cockroachdb/pebble/internal/keyspan/merging_iter.go:369 +0xc0
github.com/cockroachdb/pebble/internal/keyspan.(*BoundedIter).SeekGE(0x140002cc668, {0x14000240250?, 0x0?, 0x0?})
        /Users/radu/go/src/github.com/cockroachdb/pebble/internal/keyspan/bounded.go:104 +0x34
github.com/cockroachdb/pebble/internal/keyspan.(*DefragmentingIter).SeekGE(0x140002cc6d8, {0x14000240250, 0x1, 0x8})
        /Users/radu/go/src/github.com/cockroachdb/pebble/internal/keyspan/defragment.go:212 +0x3c
github.com/cockroachdb/pebble/internal/keyspan.(*InterleavingIter).keyspanSeekGE(0x140002cc028, {0x14000240250?, 0x0?, 0x0?}, {0x0?, 0x0?, 0x0?})
        /Users/radu/go/src/github.com/cockroachdb/pebble/internal/keyspan/interleaving_iter.go:824 +0x34
github.com/cockroachdb/pebble/internal/keyspan.(*InterleavingIter).InitSeekGE(0x140002cc028, {0x0, 0x0, 0x0}, {0x14000240250, 0x1, 0x8}, 0x0?, {{0x0?, 0x100e8458c?, ...}, ...})
        /Users/radu/go/src/github.com/cockroachdb/pebble/internal/keyspan/interleaving_iter.go:241 +0x110
github.com/cockroachdb/pebble.(*lazyCombinedIter).initCombinedIteration(0x14000498458, 0x1, 0x0, {{0x0?, 0x0?, 0x0?}, 0x0?}, {0x14000240250?, 0x140002a5b98?, 0x1009598cc?})
        /Users/radu/go/src/github.com/cockroachdb/pebble/range_keys.go:579 +0x444
github.com/cockroachdb/pebble.(*lazyCombinedIter).SeekGE(0x14000498458, {0x14000240250, 0x1, 0x8}, 0x0?)
        /Users/radu/go/src/github.com/cockroachdb/pebble/range_keys.go:592 +0x80
github.com/cockroachdb/pebble.(*Iterator).SeekGEWithLimit(0x14000498000, {0x14000240250, 0x1, 0x8}, {0x0, 0x0, 0x0})
        /Users/radu/go/src/github.com/cockroachdb/pebble/iterator.go:1319 +0x56c
github.com/cockroachdb/pebble.(*Iterator).SeekGE(...)
        /Users/radu/go/src/github.com/cockroachdb/pebble/iterator.go:1203
github.com/cockroachdb/pebble/metamorphic.(*retryableIter).SeekGE.func1()
        /Users/radu/go/src/github.com/cockroachdb/pebble/metamorphic/retryable.go:139 +0x44
github.com/cockroachdb/pebble/metamorphic.(*retryableIter).withRetry(0x14000151900, 0x140002b3d38)
        /Users/radu/go/src/github.com/cockroachdb/pebble/metamorphic/retryable.go:39 +0x34
github.com/cockroachdb/pebble/metamorphic.(*retryableIter).SeekGE(0x140002a5e88?, {0x14000240250?, 0x140002a5f38?, 0x2?})
        /Users/radu/go/src/github.com/cockroachdb/pebble/metamorphic/retryable.go:139 +0x58
github.com/cockroachdb/pebble/metamorphic.(*iterSeekGEOp).run(0x1400024c700, 0x140002a5f00?, {0x140001511a0?, 0x0?})
        /Users/radu/go/src/github.com/cockroachdb/pebble/metamorphic/ops.go:1197 +0xc0
github.com/cockroachdb/pebble/metamorphic.Execute.func1()
        /Users/radu/go/src/github.com/cockroachdb/pebble/metamorphic/meta.go:556 +0x134
golang.org/x/sync/errgroup.(*Group).Go.func1()
        /Users/radu/go/pkg/mod/golang.org/x/sync@v0.0.0-20220722155255-886fb9371eb4/errgroup/errgroup.go:75 +0x58
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 20
        /Users/radu/go/pkg/mod/golang.org/x/sync@v0.0.0-20220722155255-886fb9371eb4/errgroup/errgroup.go:72 +0x98
FAIL    github.com/cockroachdb/pebble/internal/metamorphic      0.411s
FAIL
```

--- a/metamorphic/generator.go
+++ b/metamorphic/generator.go
@@ -205,6 +205,8 @@ func generate(rng *rand.Rand, count uint64, cfg OpConfig, km *keyManager) []op {
 	}
 
 	g.dbClose()
+
+	computeDerivedFields(g.ops)
 	return g.ops
 }
 
@@ -507,7 +509,7 @@ func (g *generator) removeBatchFromGenerator(batchID objID) {
 	for _, id := range iters.sorted() {
 		g.liveIters.remove(id)
 		delete(g.iters, id)
-		g.add(&closeOp{objID: id, derivedDBID: g.objDB[batchID]})
+		g.add(&closeOp{objID: id})
 	}
 }
 
@@ -519,7 +521,7 @@ func (g *generator) batchAbort() {
 	batchID := g.liveBatches.rand(g.rng)
 	g.removeBatchFromGenerator(batchID)
 
-	g.add(&closeOp{objID: batchID, derivedDBID: g.objDB[batchID]})
+	g.add(&closeOp{objID: batchID})
 }
 
 func (g *generator) batchCommit() {
@@ -551,7 +553,7 @@ func (g *generator) batchCommit() {
 		dbID:    dbID,
 		batchID: batchID,
 	})
-	g.add(&closeOp{objID: batchID, derivedDBID: dbID})
+	g.add(&closeOp{objID: batchID})
 
 }
 
@@ -566,9 +568,8 @@ func (g *generator) dbClose() {
 	}
 	for len(g.liveBatches) > 0 {
 		batchID := g.liveBatches[0]
-		dbID := g.objDB[batchID]
 		g.removeBatchFromGenerator(batchID)
-		g.add(&closeOp{objID: batchID, derivedDBID: dbID})
+		g.add(&closeOp{objID: batchID})
 	}
 	for len(g.dbs) > 0 {
 		db := g.dbs[0]
@@ -650,9 +651,8 @@ func (g *generator) dbRestart() {
 	// Close the batches.
 	for len(g.liveBatches) > 0 {
 		batchID := g.liveBatches[0]
-		dbID := g.objDB[batchID]
 		g.removeBatchFromGenerator(batchID)
-		g.add(&closeOp{objID: batchID, derivedDBID: dbID})
+		g.add(&closeOp{objID: batchID})
 	}
 	if len(g.liveReaders) != len(g.dbs) || len(g.liveWriters) != len(g.dbs) {
 		panic(fmt.Sprintf("unexpected counts: liveReaders %d, liveWriters: %d",
@@ -846,14 +846,9 @@ func (g *generator) iterClose(iterID objID) {
 	if readerIters, ok := g.iters[iterID]; ok {
 		delete(g.iters, iterID)
 		delete(readerIters, iterID)
-		//lint:ignore SA9003 - readability
-	} else {
-		// NB: the DB object does not track its open iterators because it never
-		// closes.
 	}
 
-	readerID := g.iterReaderID[iterID]
-	g.add(&closeOp{objID: iterID, derivedDBID: g.objDB[readerID]})
+	g.add(&closeOp{objID: iterID})
 }
 
 func (g *generator) iterSetBounds(iterID objID) {
@@ -1273,10 +1268,10 @@ func (g *generator) snapshotClose() {
 	for _, id := range iters.sorted() {
 		g.liveIters.remove(id)
 		delete(g.iters, id)
-		g.add(&closeOp{objID: id, derivedDBID: g.objDB[snapID]})
+		g.add(&closeOp{objID: id})
 	}
 
-	g.add(&closeOp{objID: snapID, derivedDBID: g.objDB[snapID]})
+	g.add(&closeOp{objID: snapID})
 }
 
 func (g *generator) writerApply() {
@@ -1328,8 +1323,7 @@ func (g *generator) writerApply() {
 		batchID:  batchID,
 	})
 	g.add(&closeOp{
-		objID:       batchID,
-		derivedDBID: dbID,
+		objID: batchID,
 	})
 }
 

--- a/metamorphic/ops.go
+++ b/metamorphic/ops.go
@@ -194,8 +194,12 @@ func (o *checkpointOp) diagramKeyRanges() []pebble.KeyRange {
 
 // closeOp models a {Batch,Iterator,Snapshot}.Close operation.
 type closeOp struct {
-	objID       objID
-	derivedDBID objID
+	objID objID
+
+	// affectedObjects is the list of additional objects that are affected by this
+	// operation, and which syncObjs() must return so that we don't perform the
+	// close in parallel with other operations to affected objects.
+	affectedObjects []objID
 }
 
 func (o *closeOp) run(t *Test, h historyRecorder) {
@@ -214,17 +218,7 @@ func (o *closeOp) run(t *Test, h historyRecorder) {
 func (o *closeOp) String() string  { return fmt.Sprintf("%s.Close()", o.objID) }
 func (o *closeOp) receiver() objID { return o.objID }
 func (o *closeOp) syncObjs() objIDSlice {
-	// Synchronize on the database so that we don't close the database before
-	// all its iterators, snapshots and batches are closed.
-	// TODO(jackson): It would be nice to relax this so that Close calls can
-	// execute in parallel.
-	if o.objID.tag() == dbTag {
-		return nil
-	}
-	if o.derivedDBID != 0 {
-		return []objID{o.derivedDBID}
-	}
-	return nil
+	return o.affectedObjects
 }
 
 func (o *closeOp) keys() []*[]byte                     { return nil }
@@ -1735,6 +1729,11 @@ func (o *dbRatchetFormatMajorVersionOp) diagramKeyRanges() []pebble.KeyRange { r
 
 type dbRestartOp struct {
 	dbID objID
+
+	// affectedObjects is the list of additional objects that are affected by this
+	// operation, and which syncObjs() must return so that we don't perform the
+	// restart in parallel with other operations to affected objects.
+	affectedObjects []objID
 }
 
 func (o *dbRestartOp) run(t *Test, h historyRecorder) {
@@ -1748,7 +1747,7 @@ func (o *dbRestartOp) run(t *Test, h historyRecorder) {
 
 func (o *dbRestartOp) String() string       { return fmt.Sprintf("%s.Restart()", o.dbID) }
 func (o *dbRestartOp) receiver() objID      { return o.dbID }
-func (o *dbRestartOp) syncObjs() objIDSlice { return nil }
+func (o *dbRestartOp) syncObjs() objIDSlice { return o.affectedObjects }
 
 func (o *dbRestartOp) keys() []*[]byte                     { return nil }
 func (o *dbRestartOp) diagramKeyRanges() []pebble.KeyRange { return nil }

--- a/metamorphic/parser.go
+++ b/metamorphic/parser.go
@@ -14,6 +14,7 @@ import (
 
 	"github.com/cockroachdb/errors"
 	"github.com/cockroachdb/pebble"
+	"golang.org/x/exp/slices"
 )
 
 type methodInfo struct {
@@ -583,9 +584,29 @@ func computeDerivedFields(ops []op) {
 		case *batchCommitOp:
 			v.dbID = objToDB[v.batchID]
 		case *closeOp:
-			if derivedDBID, ok := objToDB[v.objID]; ok && v.objID.tag() != dbTag {
-				v.derivedDBID = derivedDBID
+			if v.objID.tag() == dbTag {
+				// Find all objects that use this db.
+				v.affectedObjects = nil
+				for obj, db := range objToDB {
+					if db == v.objID {
+						v.affectedObjects = append(v.affectedObjects, obj)
+					}
+				}
+				// Sort so the output is deterministic.
+				slices.Sort(v.affectedObjects)
+			} else if dbID, ok := objToDB[v.objID]; ok {
+				v.affectedObjects = objIDSlice{dbID}
 			}
+		case *dbRestartOp:
+			// Find all objects that use this db.
+			v.affectedObjects = nil
+			for obj, db := range objToDB {
+				if db == v.dbID {
+					v.affectedObjects = append(v.affectedObjects, obj)
+				}
+			}
+			// Sort so the output is deterministic.
+			slices.Sort(v.affectedObjects)
 		case *ingestOp:
 			v.derivedDBIDs = make([]objID, len(v.batchIDs))
 			for i := range v.batchIDs {

